### PR TITLE
JBIDE-21835 - fix incremental deployment by converting openshift serv…

### PR DESCRIPTION
**Pull request by robstryker**
…er to subsystems

### CODE REVIEW

**Comment by robstryker** [comment]
This also fixes the startup / publishing issue (kinda) by setting initial state to 'stopped'

**Comment by fbricon** [comment]
tests?

File: tests/org.jboss.tools.openshift.test/src/org/jboss/tools/openshift/test/core/server/OpenShiftModuleDeploymentPrefsUtilTest.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
+package org.jboss.tools.openshift.test.core.server;
```
Comment (fbricon) [inline tests/org.jboss.tools.openshift.test/src/org/jboss/tools/openshift/test/core/server/OpenShiftModuleDeploymentPrefsUtilTest.java]: missing copyright header

File: tests/org.jboss.tools.openshift.test/src/org/jboss/tools/openshift/test/core/server/OpenShiftSubsystemTest.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+ * of the various systems. 
+ * 
+ * @author rob
+ *
+ */
+public class OpenShiftSubsystemTest extends TestCase {
+
+	@After
+	public void tearDown() {
+		IServer[] all = ServerCore.getServers();
```
Comment (fbricon) [inline tests/org.jboss.tools.openshift.test/src/org/jboss/tools/openshift/test/core/server/OpenShiftSubsystemTest.java]: can you refactor that in a test utility class?

File: tests/org.jboss.tools.openshift.test/META-INF/MANIFEST.MF [hunk 3]
```diff
@@ -29,5 +29,9 @@ Require-Bundle: org.jboss.tools.openshift.client;bundle-version="[3.0.0,4.0.0)",
  org.eclipse.core.expressions;bundle-version="3.5.0",
  org.apache.commons.io;bundle-version="2.2.0",
  org.eclipse.jdt.launching;bundle-version="3.8.0",
- org.eclipse.jdt.launching.macosx;bundle-version="3.3.0";resolution:=optional
+ org.eclipse.jdt.launching.macosx;bundle-version="3.3.0";resolution:=optional,
```
Comment (fbricon) [inline tests/org.jboss.tools.openshift.test/META-INF/MANIFEST.MF]: ^ should be removed if you rebase your branch

**Comment by fbricon** [comment]
Tried to create a new Server adapter for a nodejs app deployed on the cdk, using the Server Adapter... menu in OpenShift Explorer. The wizard opens ok, but on Finish, nothing happens. The log shows:
```
java.lang.IllegalArgumentException: Add and remove cannot both be null/empty
	at org.eclipse.wst.server.core.internal.Server.canModifyModules(Server.java:2479)
	at org.eclipse.wst.server.core.internal.ServerWorkingCopy.modifyModules(ServerWorkingCopy.java:545)
	at org.jboss.tools.openshift.internal.ui.server.ServerSettingsViewModel.updateServer(ServerSettingsViewModel.java:322)
	at org.jboss.tools.openshift.internal.ui.server.ServerSettingsViewModel.updateServer(ServerSettingsViewModel.java:299)
	at org.jboss.tools.openshift.internal.ui.server.ServerSettingsWizard.performFinish(ServerSettingsWizard.java:60)
	at org.eclipse.jface.wizard.WizardDialog.finishPressed(WizardDialog.java:799)
	at org.eclipse.jface.wizard.WizardDialog.buttonPressed(WizardDialog.java:429)
```

File: plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/ServerSettingsViewModel.java [hunk 4]
```diff
@@ -301,8 +309,31 @@ private void updateServer(IServerWorkingCopy server) throws OpenShiftException {
...
+		// Set the profile
+		String profile = getProfileId();
+		ServerProfileModel.setProfile(server, profile);
+		
 		OpenShiftServerUtils.updateServerProject(
 				connectionUrl, getService(), sourcePath, podPath, routeURL, deployProject);
+		
+		IModule[] matchingModules = ServerUtil.getModules(deployProject);
+		try {
+			server.modifyModules(matchingModules, new IModule[]{}, new NullProgressMonitor());
```
Comment (fbricon) [inline plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/server/ServerSettingsViewModel.java]: this fails for nodejs apps

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/behavior/OpenShiftPublishController.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
+package org.jboss.tools.openshift.core.server.behavior;
```
Comment (fbricon) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/behavior/OpenShiftPublishController.java]: missing copyright header

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/behavior/OpenShiftEapDeployPathController.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
+package org.jboss.tools.openshift.core.server.behavior;
```
Comment (fbricon) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/behavior/OpenShiftEapDeployPathController.java]: copyright header

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/behavior/OpenShiftDeploymentOptionsController.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
+package org.jboss.tools.openshift.core.server.behavior;
```
Comment (fbricon) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/behavior/OpenShiftDeploymentOptionsController.java]: copyright header

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/behavior/UnstoppableShutdownController.java [hunk 8]
```diff
@@ -0,0 +1,18 @@
+package org.jboss.tools.openshift.core.server.behavior;
```
Comment (fbricon) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/behavior/UnstoppableShutdownController.java]: copyright header

**Comment by fbricon** [comment]
Full publish doesn't seem to create a .dodeploy marker anymore

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/behavior/OpenShiftPublishController.java [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+		} finally {
+			rsync = null;
+		}
+	}
+	
+	@Override
+	protected boolean supportsJBoss7Markers() {
+		// TODO this is the same hack as in the wizard, and should be replaced with something better
+		IService service = OpenShiftServerUtils.getService(getServer());
+		return service.getName().startsWith("eap");
```
Comment (fbricon) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/behavior/OpenShiftPublishController.java]: This works much better : 
```java
String templateName = service.getLabels().getOrDefault("template", "");
return templateName.startsWith("eap");
```

**Comment by robstryker** [comment]
"Full publish doesn't seem to create a .dodeploy marker anymore" was related to bad logic when choosing profile. It's been fixed. PR's been merged.

**Pull request merged**